spec/services/fetch_coronavirus_statistics_service_spec.rb

```diff
@@ -0,0 +1,124 @@
+        hospital_admissions_date: Date.new(2021, 3, 14),
+        new_positive_tests: 6303,
+        new_positive_tests_date: Date.new(2021, 3, 18),
+      )
+    end
+
+    context "when there is missing data" do
+      it "returns nil for no data" do
+        stub_request(:get, /coronavirus.data.gov.uk/)
+          .to_return(status: 200, body: { data: [] }.to_json)
+
+        expect(described_class.call).to be_nil
+      end
+
+      it "sets only the fields that have data" do
+        body = { data: [{ "date" => "2021-03-18",
+                          "cumulativeVaccinations" => nil,
+                          "hospitalAdmissions" => nil,
+                          "newPositiveTests" => 6303 }] }
+
+        stub_request(:get, /coronavirus.data.gov.uk/)
+          .to_return(status: 200, body: body.to_json)
+
+        expect(described_class.call).to have_attributes(
+          cumulative_vaccinations: nil,
+          cumulative_vaccinations_date: nil,
+          hospital_admissions: nil,
+          hospital_admissions_date: nil,
+          new_positive_tests: 6303,
+          new_positive_tests_date: Date.new(2021, 3, 18),
+        )
+      end
+    end
+
+    context "when the request to load data fails" do
+      it "notifies GovukError for a general error" do
+        stub_request(:get, /coronavirus.data.gov.uk/).to_return(status: 500)
+        expect(GovukError).to receive(:notify)
+
+        described_class.call
+      end
+
+      it "doens't notify GovukError when a timeout occurs" do
+        stub_request(:get, /coronavirus.data.gov.uk/).to_timeout
+        expect(GovukError).not_to receive(:notify)
+
+        described_class.call
+      end
+
+      it "doens't notify GovukError on 502 and 504 errors" do
+        stub_request(:get, /coronavirus.data.gov.uk/)
+          .to_return(status: 502)
+          .then
+          .to_return(status: 504)
+
+        expect(GovukError).not_to receive(:notify)
+
+        2.times { described_class.call }
+      end
+    end
+
+    context "when the cache has stale data and the request to load data fails" do
+      before do
+        stub_request(:get, /coronavirus.data.gov.uk/).to_return(status: 500)
+      end
+
+      it "returns Statistics with the stale data" do
+        stale_stats = {
+          cumulative_vaccinations: 25_735_472,
+          cumulative_vaccinations_date: Date.new(2021, 3, 17),
+          hospital_admissions: 426,
+          hospital_admissions_date: Date.new(2021, 3, 14),
+          new_positive_tests: 6303,
+          new_positive_tests_date: Date.new(2021, 3, 18),
+        }
+
+        allow(Rails.cache)
+          .to receive(:fetch)
+          .with(described_class::CACHE_KEY)
+          .and_return(stale_stats)
+
+        allow(Rails.cache)
+          .to receive(:fetch)
+          .with(described_class::CACHE_KEY, expires_in: 30.minutes, race_condition_ttl: 5.minutes)
+          .and_call_original
+
+        expect(described_class.call).to have_attributes(stale_stats)
```

Comment:
Sorry, I'm a bit confused, how does this test work? How do we know that cache data is stale?

Reply:
Yeah, this is a bit nuanced - it was quite hard to document as a test. I've refactored it a little to try make it a better read, though I don't think it's ever going to be great.

So the code works by first making a request to the cache without an expiry to load data whether it's fresh or stale, and then it then makes a second call to the cache which will load fresh data or make the API request.

What this test does is it stubs the call of the first cache call, so the output of that is predictable. It then has the second cache call fail. By the code returning the data from the first cache call we know that it had to have been from the first cache and therefore resilient to failure on the second call.